src/feditest/nodedrivers/imp/__init__.py

```diff
@@ -62,60 +62,61 @@ def http(self, request: HttpRequest, follow_redirects: bool = True) -> HttpReque
         first_request = HttpRequest(parsed_uri)
         current_request = first_request
         pair : HttpRequestResponsePair | None = None
         for redirect_count in range(10, 0, -1):
             pair = self.http(current_request)
             if pair.response and pair.response.is_redirect():
                 if redirect_count <= 0:
-                    raise WebClient.TooManyRedirectsError(current_request)
+                    return WebFingerQueryResponse(pair, None, WebClient.TooManyRedirectsError(current_request))
                 parsed_location_uri = ParsedUri.parse(pair.response.location())
                 if not parsed_location_uri:
-                    raise ValueError('Location header is not a valid URI:', uri, '(from', resource_uri, ')')
+                    return WebFingerQueryResponse(pair, None, ValueError('Location header is not a valid URI:', uri, '(from', resource_uri, ')'))
                 current_request = HttpRequest(parsed_location_uri)
             break
 
-        # I guess we always have a non-null responses here, but mypy complains without the if
-        if pair:
-            ret_pair = HttpRequestResponsePair(first_request, current_request, pair.response)
-            if ret_pair.response is not None:
-                if ret_pair.response.http_status == 200:
-                    if (
-                        not check_content_type
-                        or ret_pair.response.content_type() == "application/jrd+json"
-                        or ret_pair.response.content_type().startswith(
-                            "application/jrd+json;"
-                        )
-                    ):
-                        if ret_pair.response.payload is not None:
-                            json_string = ret_pair.response.payload.decode(
-                                encoding=ret_pair.response.payload_charset() or "utf8" )
-                            jrd = ClaimedJrd(json_string)
-                            if validate_jrd:
-                                try:
-                                    jrd.validate()
-                                except Exception as ex:
-                                    raise AssertionError(*ex.args[1:])
-                            return WebFingerQueryResponse(pair, jrd)
-                        raise WebFingerClient.WebfingerQueryFailedError(
-                            uri, ret_pair, "No payload"
-                        )
-                    raise WebFingerClient.WebfingerQueryFailedError(uri, ret_pair, f"Invalid content type: { ret_pair.response.content_type() }")
-                raise WebFingerClient.WebfingerQueryFailedError(uri, ret_pair, f"Invalid HTTP status: { ret_pair.response.http_status }")
-
-        raise WebFingerClient.WebfingerQueryFailedError(uri, ret_pair)
+        # I guess we always have a non-null responses here, but mypy complains without the cast
+        pair = cast(HttpRequestResponsePair, pair)
```

Comment:
You declare `pair` to be nullable but `self.http` is not declared to return a nullable value. If `self.http` raises an exception, the previously-conditional code doesn't run anyway. I might be missing something, but maybe it would be cleaner to refactor the content type and JRD validations into a separate method and call that from the redirect loop when you have an otherwise valid response. The `cast` shouldn't be necessary with that approach.

Reply:
Feel free :-)